pkg/certificate/signing_requestor.go

```diff
@@ -0,0 +1,487 @@
+
+// startCertificateRenewalMonitoring starts periodic monitoring of certificate expiration
+func (s *signingRequestorImpl) startCertificateRenewalMonitoring(stopCh <-chan struct{}) {
+	s.renewalTicker = time.NewTicker(CertCheckInterval)
+	s.renewalStopCh = make(chan struct{})
+
+	go func() {
+		defer s.renewalTicker.Stop()
+
+		for {
+			select {
+			case <-s.renewalTicker.C:
+				s.checkCertificateRenewal()
+			case <-s.renewalStopCh:
+				return
+			case <-stopCh:
+				return
+			}
+		}
+	}()
+
+	logger.Info("Started certificate renewal monitoring", "interval", CertCheckInterval)
+}
+
+// stopCertificateRenewalMonitoring stops the certificate renewal monitoring
+func (s *signingRequestorImpl) stopCertificateRenewalMonitoring() {
+	if s.renewalTicker != nil {
+		s.renewalTicker.Stop()
+		s.renewalTicker = nil
+	}
+
+	if s.renewalStopCh != nil {
+		close(s.renewalStopCh)
+		s.renewalStopCh = nil
+	}
+
+	logger.Info("Stopped certificate renewal monitoring")
+}
+
+// updateCertificateExpiration extracts and stores the certificate expiration time
+func (s *signingRequestorImpl) updateCertificateExpiration(secret *corev1.Secret) {
+	certPEM := secret.Data[TLSDataKey]
+	if len(certPEM) == 0 {
+		return
+	}
+
+	block, _ := pem.Decode(certPEM)
+	if block == nil {
+		logger.Warningf("Failed to decode certificate PEM for secret %q", secret.Name)
+		return
+	}
+
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		logger.Errorf(err, "Failed to parse certificate for secret %q", secret.Name)
+		return
+	}
+
+	// Update the expiration time in our tracking map
+	if v, ok := s.issuedCerts.Load(secret.Name); ok {
+		renewalInfo := v.(certRenewalInfo)
+		renewalInfo.expiresAt = cert.NotAfter
+		s.issuedCerts.Store(secret.Name, renewalInfo)
+
+		logger.V(1).Infof("Updated certificate expiration for %q: expires at %s",
+			secret.Name, cert.NotAfter.Format(time.RFC3339))
+	}
+}
+
+// checkCertificateRenewal checks all tracked certificates and renews those close to expiration
+func (s *signingRequestorImpl) checkCertificateRenewal() {
+	s.issuedCerts.Range(func(key, value interface{}) bool {
+		secretName := key.(string)
+		renewalInfo := value.(certRenewalInfo)
+
+		// Skip if expiration time is not set yet
+		if renewalInfo.expiresAt.IsZero() {
+			return true
+		}
+
+		// Check if certificate needs renewal
+		timeUntilExpiry := time.Until(renewalInfo.expiresAt)
+		if timeUntilExpiry <= CertRenewBefore {
+			logger.Infof("Certificate %q expires in %s, renewing", secretName, timeUntilExpiry.String())
+
+			// Renew the certificate by re-issuing it
+			if err := s.Issue(context.TODO(), renewalInfo.name, renewalInfo.ips, renewalInfo.onSigned); err != nil {
+				logger.Errorf(err, "Failed to renew certificate %q", secretName)
+			} else {
+				logger.Infof("Successfully initiated renewal for certificate %q", secretName)
+			}
+		} else {
+			logger.V(2).Infof("Certificate %q expires in %s, no renewal needed", secretName, timeUntilExpiry.String())
```

Comment:
We should use our log constants. I think this should be TRACE.

```suggestion
			logger.V(log.TRACE).Infof("Certificate %q expires in %s, no renewal needed", secretName, timeUntilExpiry.String())
```